minio/credentials/providers.py

```diff
@@ -0,0 +1,238 @@
+# -*- coding: utf-8 -*-
+# MinIO Python Library for Amazon S3 Compatible Cloud Storage,
+# (C) 2020 MinIO, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Credential providers."""
+
+import configparser
+import json
+import os
+import sys
+from datetime import datetime, timedelta
+
+import urllib3
+from urllib3.exceptions import HTTPError, ResponseError
+
+from .credentials import Provider, Value
+
+RFC3339NANO = "%Y-%m-%dT%H:%M:%S.%fZ"
+RFC3339 = "%Y-%m-%dT%H:%M:%SZ"
+
+
+class AssumeRoleProvider(Provider):
+    """Assume-role credential provider."""
+
+    def __init__(self, provider_func):
+        self._provider_func = provider_func
+        self._value = None
+        self._expiry = None
+
+    def retrieve(self):
+        if (
+                not self._value or
+                (self._expiry and self._expiry < datetime.utcnow())
+        ):
+            self._value, self._expiry = self._provider_func()
+        return self._value, self._expiry
+
+
+class Chain(Provider):
+    """Chained credential provider."""
+
+    def __init__(self, providers):
+        self._providers = providers
+        self._provider = None
+
+    def retrieve(self):
+        try:
+            if self._provider:
+                return self._provider.retrieve()
+        except ValueError:
+            # ignore this error and try other providers.
+            pass
+
+        for provider in self._providers:
+            try:
+                creds, expiry = provider.retrieve()
+                if creds:
+                    self._provider = provider
+                    return creds, expiry
+            except ValueError:
+                # ignore this error and try other providers.
+                pass
+
+        raise ValueError("no credentials retrieved")
```

Comment:
Maybe here we should return the last encountered ValueError exception.. so at least the user can see at least a more self explaining error.

Reply:
Done